subprojects/base-services/src/main/java/org/gradle/internal/hash/Hashing.java

```diff
@@ -117,109 +118,76 @@ public static HashFunction sha1() {
         return SHA1;
     }
 
-    private static abstract class MessageDigestHashFunction implements HashFunction {
-        public static MessageDigestHashFunction of(String algorithm) {
-            MessageDigest prototype;
-            try {
-                prototype = MessageDigest.getInstance(algorithm);
-            } catch (NoSuchAlgorithmException e) {
-                throw new IllegalArgumentException("Cannot instantiate digest algorithm: " + algorithm);
-            }
-            try {
-                prototype.clone();
-                return new CloningMessageDigestHashFunction(prototype);
-            } catch (CloneNotSupportedException e) {
-                return new RegularMessageDigestHashFunction(algorithm);
-            }
+    private static class MessageDigestHashFunction implements HashFunction {
+        public static MessageDigestHashFunction of(final String algorithm) {
+            ThreadLocal<MessageDigest> messageDigestThreadLocal = ThreadLocal.withInitial(new Supplier<MessageDigest>() {
```

Comment:
`ThreadLocal` with lazy init per thread instead of `MessageDigest` clone

Reply:
This will cause a memory leak if the ThreadLocal is never removed.

Reply:
Thanks Stefan, when do you think it would be appropriate to remove it?

Reply:
I don't yet understand the Gradle architecture well enough to be able to tell if this is a problem.

This hold a JDK class, so not sure how it can cause a leak.

Until now this was held statically a single time (and cloned for every usage), now every thread can have a copy of a message digest. Is that a problem?

Could you please point me to the problematic code so that I can check it?

Reply:
The problem with ThreadLocals is that they stay in memory until they are either removed or the thread is stopped. You are right that a JVM type is less of a problem, because it least it won't lead to a classloader leak. It will still leak the MessageDigest though. It's not a big thing, more a matter of design principle. 

I'd be fine with it if we can demonstrate a measurable improvement in one of our performance tests. Microbenchmarks don't count, since they don't reflect how Gradle is really used :)

Reply:
I agree, though I'm not sure if the performance tests can catch heap reduction.
I understood from @lptr that the perf tests are not stable at the moment, will get back to this, once they can be run reliably again.

Reply:
The performance tests would catch improved performance due to less GC.